File: cppsrc/Network.cpp
#include "settings.h"
#include "Network.h"
#include <iostream>
#include <fstream>
#include <sstream>
#include <cmath>

using namespace copilot;

    Network::Network() {
        std::vector<std::vector<std::vector<double>>> weights = {};
        std::vector<std::vector<double>> biases = {};
        std::vector<std::vector<double>> weighted_sums = {};
        std::vector<std::vector<double>> neurons = {};
        int layers = Settings().layers;
        std::string activation = Settings().activation;
    }

    Network::~Network(){

    }


    void Network::load_from_file(std::string f_weights) {
        int i = 0;
        std::ifstream file(f_weights);
        std::string line;

        while (std::getline(file, line)) {
            std::vector<double> data;
            std::stringstream ss(line);
            std::string item;

            while (std::getline(ss, item, ',')) {
                data.push_back(std::stod(item));
            }

            if (i % 2 == 0) {
                // weights
                std::vector<std::vector<double>> new_weight;
                std::vector<double> new_row;

                for (double num : data) {
                    new_row.push_back(num);
                    if (new_row.size() >= Settings().layout[i / 2]) {
                        new_weight.push_back(new_row);
                        new_row.clear();
                    }
                }

                // check if rows and columns match up
                if (new_weight.size() != Settings().layout[i / 2 + 1] || new_weight[0].size() != Settings().layout[i / 2]) {
                    std::cout << "the weights at line " << i << " do not match up with settings" << std::endl;
                    return;
                }

                weights.push_back(new_weight);
            } else {
                // biases
                std::vector<double> new_biases;

                for (double num : data) {
                    new_biases.push_back(num);
                }

                // check if vector length match up
                if (new_biases.size() != Settings().layout[i / 2 + 1]) {
                    std::cout << "the biases at line " << i << " do not match up with settings" << std::endl;
                    return;
                }

                biases.push_back(new_biases);
            }

            i++;
        }

        file.close();
    }

    void Network::save(std::string outfile, const std::vector<std::vector<std::vector<double>>>& weights, const std::vector<std::vector<double>>& biases) {
        std::ofstream file(outfile);

        if (!file.is_open()) {
            std::cout << "Failed to open the file." << std::endl;
            return;
        }

        for (int i = 0; i < weights.size(); i++) {
            std::stringstream ss;

            for (const auto& row : weights[i]) {
                for (double num : row) {
                    ss << num << ", ";
                }
                ss << "\n";
            }

            file << ss.str();

            for (double bias : biases[i]) {
                file << bias << ", ";
            }
            file << "\n";
        }

        file.close();
    }

    std::vector<double> Network::feed(std::vector<double> input) {
        weighted_sums.clear();
        neurons.clear();
        return forward_propagate(input, 0);
    }

    std::vector<double> Network::forward_propagate(std::vector<double> input, int layer) {
            this->neurons.push_back(input);
            if (layer >= this->layers - 1) {
                return input;
            }
            std::vector<double> weighted_sums;
            for (int i = 0; i < this->weights[layer].size(); i++) {
                double neuron_weighted_sum = 0;
                for (int j = 0; j < this->weights[layer][i].size(); j++) {
                    neuron_weighted_sum += this->weights[layer][i][j] * input[j];
                }
                neuron_weighted_sum += this->biases[layer][i];
                weighted_sums.push_back(neuron_weighted_sum);
            }
            this->weighted_sums.push_back(weighted_sums);
            std::vector<double> next_layer;
            if (layer >= this->layers - 2) {
                next_layer = softmax(weighted_sums);
            } else if (this->activation == "relu") {
                next_layer = ReLU_vectorized(weighted_sums);
            } else if (this->activation == "sigmoid") {
                next_layer = sigmoid_vectorized(weighted_sums);
            } else if (this->activation == "tanh") {
                next_layer = tanh_vectorized(weighted_sums);
            }
            return forward_propagate(next_layer, layer + 1);
    }
    std::vector<double> ReLU_vectorized(std::vector<double> x) {
        std::vector<double> result;
        for (int i = 0; i < x.size(); i++) {
            result.push_back(x[i] * (x[i] > 0));
        }
        return result;
    }
    std::vector<double> dReLU_vectorized(std::vector<double> x) {
        std::vector<double> result;
        for (int i = 0; i < x.size(); i++) {
            result.push_back(1. * (x[i] > 0));
        }
        return result;
    }
    std::vector<double> sigmoid_vectorized(std::vector<double> x) {
        std::vector<double> result;
        for (int i = 0; i < x.size(); i++) {
            result.push_back(1 / (1 + exp(-x[i])));
        }
        return result;
    }
    std::vector<double> dsigmoid_vectorized(std::vector<double> x) {
        std::vector<double> result;
        for (int i = 0; i < x.size(); i++) {
            result.push_back((1/(1+exp(-(1*(x[i]))))) * (1 - (1/(1+exp(-(1*(x[i])))))));
        }
        return result;
    }
    std::vector<double> softmax(std::vector<double> x) {
        std::vector<double> result;
        double sum = 0;
        for (int i = 0; i < x.size(); i++) {
            sum += exp(x[i]);
        }
        for (int i = 0; i < x.size(); i++) {
            result.push_back(exp(x[i]) / sum);
        }
        return result;
    }
    std::vector<double> tanh_vectorized(std::vector<double> x) {
        std::vector<double> result;
        for (int i = 0; i < x.size(); i++) {
            result.push_back(tanh(x[i]));
        }
        return result;
    }
    std::vector<double> dtanh_vectorized(std::vector<double> x) {
        std::vector<double> result;
        for (int i = 0; i < x.size(); i++) {
            result.push_back(1 - pow(tanh(x[i]), 2));
        }
        return result;
    }

File: pysrc/Network.py
import numpy as np
from settings import *

class Network:
    def __init__(self, *args):
        self.layers = Settings.layers
        self.activation = Settings.activation
        self.weights = []
        self.biases = []
        self.weighted_sums = []
        self.neurons = []

        if len(args) == 1:
            self.load_from_file(args[0])
        elif len(args) == 2:
            self.weights = args[0]
            self.biases = args[1]
        
    
    def load_from_file(self, f_weights: str):
        i = 0
        for line in open(f_weights, "r"):
            data = [float(x) for x in line.split(", ")]
            if (i % 2 == 0):
                # weights
                new_weight = []
                new_row = []
                for num in data:
                    new_row.append(num)
                    if len(new_row) >= Settings.layout[i//2]:
                        new_weight.append(new_row)
                        new_row = []

                new_weight = np.array(new_weight)

                # check if rows and columns match up
                if new_weight.shape != (Settings.layout[i//2+1], Settings.layout[i//2]):
                    print(f"the weights at line {i} do not match up with settings")
                    return
                    
                self.weights.append(new_weight)   
                
            else:
                # biases
                new_biases = []
                for num in data:
                    new_biases.append(num)
                new_biases = np.array(new_biases)

                # check if vector length match up
                if new_biases.shape != (Settings.layout[i//2+1],):
                    print(f"the biases at line {i} do not match up with settings")
                    return

                self.biases.append(new_biases)

            i += 1
    
    # serialize weights and biases into text file
    def save(self, outfile: str):
        file = open(outfile, "w")
        
        for i in range(0, self.layers - 1):
            weights = ", ".join(", ".join(str(y) for y in x) for x in self.weights[i])
            file.write(weights + "\n")
            biases = ", ".join(str(x) for x in self.biases[i])
            file.write(biases + "\n")
        
        file.close()
    
    # recursive feed
    def feed(self, input: np.ndarray) -> np.ndarray:
        self.weighted_sums = []
        self.neurons = []

        return self.forward_propagate(input, 0)
    
    def forward_propagate(self, input: np.ndarray, layer: int) -> np.ndarray:
        self.neurons.append(input)

        if layer >= self.layers - 1:
            return input
            
        # the sum of weighted activations plus bias
        weighted_sums = np.dot(self.weights[layer], input) + self.biases[layer][:, np.newaxis]
        # if second to last layer, then apply softmax instead of activation func
        if layer >= self.layers - 2:
            next_layer = softmax(weighted_sums) #only for use in multi-class classification problems
        elif self.activation == "relu":
            next_layer = np.vectorize(ReLU)(weighted_sums)
        elif self.activation == "sigmoid":
            next_layer = np.vectorize(sigmoid)(weighted_sums)
        elif self.activation == "tanh":
            next_layer = np.vectorize(tanh)(weighted_sums)
        
        self.weighted_sums.append(weighted_sums)
        return self.forward_propagate(next_layer, layer + 1)

            

def ReLU(x):
    return x * (x > 0)

def dReLU(x):
    return 1. * (x > 0)

def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def dsigmoid(x):
    return sigmoid(x) * (1 - sigmoid(x))

def softmax(x):
    return np.exp(x) / np.sum(np.exp(x), axis=0)

def tanh(x):
    return np.tanh(x)

def dtanh(x):
    return 1 - tanh(x)**2
